HeterogeneousCore/SonicTriton/src/TritonClient.cc

```diff
@@ -0,0 +1,318 @@
+#include "FWCore/Utilities/interface/Exception.h"
+#include "HeterogeneousCore/SonicTriton/interface/TritonClient.h"
+
+#include "request_grpc.h"
+
+#include <string>
+#include <cmath>
+#include <chrono>
+#include <exception>
+#include <sstream>
+#include <cstring>
+
+namespace ni = nvidia::inferenceserver;
+namespace nic = ni::client;
+
+//based on https://github.com/NVIDIA/triton-inference-server/blob/v1.12.0/src/clients/c++/examples/simple_callback_client.cc
+
+template <typename Client>
+TritonClient<Client>::TritonClient(const edm::ParameterSet& params)
+    : Client(),
+      url_(params.getParameter<std::string>("address") + ":" + std::to_string(params.getParameter<unsigned>("port"))),
+      timeout_(params.getParameter<unsigned>("timeout")),
+      modelName_(params.getParameter<std::string>("modelName")),
+      modelVersion_(params.getParameter<int>("modelVersion")),
+      batchSize_(params.getParameter<unsigned>("batchSize")),
+      nInput_(params.getParameter<unsigned>("nInput")),
+      nOutput_(params.getParameter<unsigned>("nOutput")),
+      verbose_(params.getParameter<bool>("verbose")),
+      allowedTries_(params.getParameter<unsigned>("allowedTries")) {
+  this->clientName_ = "TritonClient";
+}
+
+template <typename Client>
+bool TritonClient<Client>::wrap(const nic::Error& err, const std::string& msg) const {
+  if (!err.IsOk()) {
+    //emit warning
+    edm::LogWarning("TritonServerWarning") << msg << ": " << err;
+  }
+  return err.IsOk();
+}
+
+template <typename Client>
+bool TritonClient<Client>::setup() {
+  bool status = true;
+
+  status = wrap(nic::InferGrpcContext::Create(&context_, url_, modelName_, modelVersion_, false),
+                "setup(): unable to create inference context");
+  if (!status)
+    return status;
+
+  //only used for monitoring
+  bool has_server = false;
+  if (verbose_) {
+    bool has_server = wrap(nic::ServerStatusGrpcContext::Create(&serverCtx_, url_, false),
+                           "setup(): unable to create server context");
+  }
+  if (!has_server)
+    serverCtx_ = nullptr;
+
+  std::unique_ptr<nic::InferContext::Options> options;
+  status = wrap(nic::InferContext::Options::Create(&options), "setup(): unable to create inference context options");
+  if (!status)
+    return status;
+
+  options->SetBatchSize(batchSize_);
+  for (const auto& output : context_->Outputs()) {
+    status = wrap(options->AddRawResult(output), "setup(): unable to add raw result");
+    if (!status)
+      return status;
+  }
+  status = wrap(context_->SetRunOptions(*options), "setup(): unable to set run options");
+  if (!status)
+    return status;
+
+  const auto& nicInputs = context_->Inputs();
+  nicInput_ = nicInputs[0];
+  nicInput_->Reset();
+
+  auto t1 = std::chrono::high_resolution_clock::now();
+  std::vector<int64_t> input_shape;
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    float* arr = &(this->input_.data()[i0 * nInput_]);

```

Comment:
Why go via `vector::data()`? Wouldn't `operator[]()` suffice?
```suggestion
    float* arr = &(this->input_[i0 * nInput_]);
```